### PR TITLE
[RLlib] Memory leak finding toolset using tracemalloc + CI/nightly memory leak tests.

### PR DESCRIPTION
**Pull request by sven1977**
Adds memory leak CI-tests to RLlib.
- Adds script and utilities: utils/tests/run_memory_leak_test.py utils/debug/memory.py
- Adds tuned_examples used for running these memory leak tests.
- Uses tracemalloc to pinpoint exactly where an allocation happened (file+line+stacktrace) that is likely never cleaned up/garbage collected.
- Simple tools for testing an entire Trainer (its sub-components: env, policy, rollout_worker, etc..).

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

File: rllib/utils/pre_checks/env.py [hunk 1]
```diff
@@ -40,35 +41,48 @@ def check_env(env: EnvType) -> None:
...
+        elif isinstance(env, BaseEnv):
+            check_base_env(env)
+        else:
+            logger.warning(
+                "Env checking isn't implemented for VectorEnvs, RemoteBaseEnvs, "
+                "ExternalMultiAgentEnv,or ExternalEnvs or Environments that are "
+                "Ray actors"
+            )
+    except Exception:
+        actual_error = traceback.format_exc()
```
Comment (gjoliver) [inline rllib/utils/pre_checks/env.py:76]: nice!
are these diffs actually part of this PR, or just needs rebase?
I feel like these env checking improvements may be worth a separate PR by themselves.

File: rllib/BUILD [hunk 2]
```diff
@@ -833,6 +843,82 @@ py_test(
...
+# Tag: memory_leak_tests
+# --------------------------------------------------------------------
+
+py_test(
+    name = "test_memory_leak_a3c",
+    tags = ["team:ml", "memory_leak_tests"],
+    main = "tests/run_memory_leak_tests.py",
+    size = "large",
+    srcs = ["tests/run_memory_leak_tests.py"],
+    data = ["tuned_examples/a3c/memory-leak-test-a3c.yaml"],
```
Comment (gjoliver) [inline rllib/BUILD:858]: another way to organize these may be to group these memory-leak-test-xxx.yaml files into their own dir under rllib/tests/memory_leak_tests/.
that way they are not part of the tuned_example folder, since they are mostly for testing purpose, not demonstration purpose.
just a suggestion.

File: rllib/tests/run_memory_leak_tests.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        print(yaml.dump(experiment))
+
+        # Construct the trainer instance based on the given config.
+        leaking = True
+        try:
+            ray.init(num_cpus=5, local_mode=args.local_mode)
+            trainer = get_trainer_class(experiment["run"])(experiment["config"])
+            results = check_memory_leaks(
+                trainer,
+                to_check=set(args.to_check) or {"rollout_worker", "policy"},
```
Comment (gjoliver) [inline rllib/tests/run_memory_leak_tests.py:117]: why not check model too?

File: rllib/utils/debug/memory.py [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+Suspect = namedtuple(
+    "Suspect",
+    [
+        # The stack trace of the allocation, going back n frames, depending
+        # on the tracemalloc.start(n) call.
+        "traceback",
+        # The amount of memory taken by this particular stack trace
+        # over the course of the experiment.
+        "memory_increase",
+        # The slope of the scipy linear regression (x=iteration; y=memory size).
```
Comment (gjoliver) [inline rllib/utils/debug/memory.py:23]: update comment? why scipy? same below.

File: rllib/utils/debug/memory.py [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+
+        test = _test_some_code_for_memory_leaks(
+            desc="Looking for leaks in env, running through episodes.",
+            init=None,
+            code=code,
+            # How many times to repeat the function call?
+            repeats=repeats or 200,
+            max_num_trials=max_num_trials,
+        )
+        if test:
```
Comment (gjoliver) [inline rllib/utils/debug/memory.py:95]: instead of early return, should we do all the tests, and union all the returned results?

File: rllib/utils/debug/memory.py [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+
+        table = defaultdict(list)
+
+        # Repeat running code for n times.
+        # Increase repeat value with each trial to make sure stats are more
+        # solid each time (avoiding false positives).
+        actual_repeats = repeats * (trial + 1)
+
+        print(desc, end="")
+        print(f" {actual_repeats} times.")
```
Comment (gjoliver) [inline rllib/utils/debug/memory.py:233]: print(f"{desc} {actual_repeats} times.")
?

File: rllib/utils/debug/memory.py [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+        suspicious_stats.clear()
+        # Suspicious memory allocation found?
+        suspects = _find_memory_leaks_in_table(table)
+        for suspect in sorted(suspects, key=lambda s: s.memory_increase, reverse=True):
+            # Only print out the biggest offender:
+            if len(suspicious) == 0:
+                _pprint_suspect(suspect)
+                print("-> added to retry list")
+            suspicious.add(suspect.traceback)
+            suspicious_stats.append(suspect)
```
Comment (gjoliver) [inline rllib/utils/debug/memory.py:256]: I am a bit confused.
are we simply appending suspects from all trials?
how are we making sure a suspect is real if we simply Union all the suspects here?

File: rllib/utils/debug/memory.py [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+            # Do a linear regression to get the slope and R-value.
+            line = scipy.stats.linregress(x=np.arange(len(hist)), y=np.array(hist))
+
+            # - If weak positive slope and some confidence and
+            #   increase > n bytes -> error.
+            # - If stronger positive slope -> error.
+            if memory_increase > 1000 and (
+                (line.slope > 20.0 and line.rvalue > 0.4)
+                or (line.slope > 10.0 and line.rvalue > 0.7)
+                or (line.slope > 5.0 and line.rvalue > 0.85)
```
Comment (gjoliver) [inline rllib/utils/debug/memory.py:334]: cool!

**Review by gjoliver — COMMENTED** [review]
really cool change. some coding suggestions.

File: rllib/env/multi_agent_env.py [hunk 9]
```diff
@@ -184,7 +184,6 @@ def action_space_sample(self, agent_ids: list = None) -> MultiAgentDict:
                 if agent_id != "__all__"
             }
         logger.warning("action_space_sample() has not been implemented")
-        del agent_ids
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/env/multi_agent_env.py:187]: No, but they don't seem to make any sense here. We shouldn't get into the habit of placing these `del`s everywhere in the code.

File: rllib/tests/run_memory_leak_tests.py [hunk 10]
```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env python
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/tests/run_memory_leak_tests.py:1]: Sure, we should probably get rid of `rllib/tests/` altogether and move all remaining tests in there into their respective sub-folders.

[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/tests/run_memory_leak_tests.py:1]: done

File: rllib/utils/pre_checks/env.py (hunk 1 above)
Comment (sven1977) [reply rllib/utils/pre_checks/env.py:76]: Yeah, I think I merged another PR into this one (to have some important fixes in place so I can continue testing). But yeah, we should merge the other PR first. ...

File: rllib/utils/debug/memory.py (hunk 6 above)
Comment (sven1977) [reply rllib/utils/debug/memory.py:233]: done

File: rllib/utils/debug/memory.py (hunk 5 above)
Comment (sven1977) [reply rllib/utils/debug/memory.py:95]: done

File: rllib/utils/debug/memory.py [hunk 11]
```diff
@@ -0,0 +1,358 @@
...
+            suspicious.add(suspect.traceback)
+            suspicious_stats.append(suspect)
+
+        tracemalloc.stop()
+
+        # Some suspicious memory allocations found.
+        if len(suspicious) > 0:
+            print(f"{len(suspicious)} suspects found. Top-ten:")
+            for i, s in enumerate(
+                sorted(suspicious_stats, key=lambda s: s.memory_increase, reverse=True)
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/utils/debug/memory.py:264]: True, fixed.

File: rllib/utils/debug/memory.py (hunk 7 above)
Comment (sven1977) [reply rllib/utils/debug/memory.py:256]: Yeah, maybe not clearly enough documented here:

* 1st pass: Store all (top 1000) suspicious traces in `table`. See `_take_snapshot()` function.
* 2nd pass: Store all (top 100) suspicious traces in table (no matter what) AND those at position 101 to 1000 that were already suspicious in 1st pass.
* 3rd pass and all others: Like 2nd pass.

"top n" here means: sorted by total memory consumption over the code execution repeats.

File: rllib/env/multi_agent_env.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/env/multi_agent_env.py:187]: oh I remember this now.
yeah, the only related case I can think of is when a function argument is unused. google's python style guide ask you to del them explicitly first thing at the start of the function.
but that's clearly not the case here.

File: .buildkite/pipeline.ml.yml [hunk 12]
```diff
@@ -101,6 +101,28 @@
...
+  conditions: ["RAY_CI_RLLIB_AFFECTED"]
+  commands:
+    - cleanup() { if [ "${BUILDKITE_PULL_REQUEST}" = "false" ]; then ./ci/travis/upload_build_info.sh; fi }; trap cleanup EXIT
+    - RLLIB_TESTING=1 PYTHON=3.7 ./ci/travis/install-dependencies.sh
+    - bazel test --config=ci $(./scripts/bazel_export_options)
+      --build_tests_only
+      --test_tag_filters=memory_leak_tests,-flaky
+      --test_arg=--framework=torch
+      rllib/...
+
```
Comment (gjoliver) [inline .buildkite/pipeline.ml.yml:125]: I was told that our CI cost has been pretty high. and there are talks about optimizing things a bit.
should we make these weekly tests, instead of pre-merge?

File: rllib/agents/impala/vtrace_tf.py [hunk 13]
```diff
@@ -360,15 +360,6 @@ def from_importance_weights(
         rhos = tf.math.exp(log_rhos)
         if clip_rho_threshold is not None:
             clipped_rhos = tf.minimum(clip_rho_threshold, rhos, name="clipped_rhos")
-
-            tf1.summary.histogram("clipped_rhos_1000", tf.minimum(1000.0, rhos))
```
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/agents/impala/vtrace_tf.py:364]: ?

File: rllib/env/vector_env.py [hunk 14]
```diff
@@ -237,11 +237,6 @@ def vector_step(self, actions):
         obs_batch, rew_batch, done_batch, info_batch = [], [], [], []
         for i in range(self.num_envs):
             obs, r, done, info = self.envs[i].step(actions[i])
-            if not np.isscalar(r) or not np.isreal(r) or not np.isfinite(r):
```
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/env/vector_env.py:240]: ?

**Review by gjoliver — COMMENTED** [review]
these new tests are still failing.
I wonder how stable these tests would be.
like what if we have a buffer that is larger than 600? there will be organic growth during the beginning period?
worth considering making these manual tests? just a question.

**Comment by sven1977** [comment]
I think that tf2 still has a leak, actually, and that's why its new memory leak tests are failing:

```
File "/ray/python/ray/rllib/policy/eager_tf_policy.py", line 177
--
  | **kwargs,
  | File "/ray/python/ray/rllib/policy/eager_tf_policy.py", line 488
  | self.global_timestep += tree.flatten(ret[0])[0].shape.as_list()[0]
  | Increase total=838600B
  | Slope=1400.0 B/detection
```

File: rllib/agents/impala/vtrace_tf.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/agents/impala/vtrace_tf.py:364]: Yes, these are memory leaking lines of code. If we want these in our stats, we should move this into the `stats_fn`.

File: rllib/env/vector_env.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/env/vector_env.py:240]: Yes and no. :) We are already checking this now via the new env checking module before.
Just wanted to remove redundancy here. Otherwise one could argue that we need to add this check to all our other env types as well.

File: .buildkite/pipeline.ml.yml (hunk 12 above)
Comment (sven1977) [reply .buildkite/pipeline.ml.yml:125]: Yeah, not sure. Maybe we should do this.
How about - as a compromise - we remove the tf1.x learning tests first? Lemme do this. ...

Comment (sven1977) [reply .buildkite/pipeline.ml.yml:112]: Done: tf1.x learning tests removed.

File: .buildkite/pipeline.ml.yml [hunk 15]
```diff
@@ -46,27 +46,7 @@
       --test_arg=--framework=tf2
       rllib/...
 
-- label: ":brain: RLlib: Learning discr. actions TF1-static-graph (from rllib/tuned_examples/*.yaml)"
```
Comment (sven1977) [inline .buildkite/pipeline.ml.yml:49]: As we move more and more heavy testing in, we should also sometimes remove some tests: E.g. tf1.x should no longer be tested, imo.

Comment (gjoliver) [reply .buildkite/pipeline.ml.yml:49]: do we want to double check with the bigger ML team on this?
we should have a coherent story about tf1 deprecation.

Comment (sven1977) [reply .buildkite/pipeline.ml.yml:49]: Yes, we should do this: But note that these are only tf1.x tests, so the tf version here is really 1.x. We are NOT disabling framework="tf" (tf static graph) tests here. Sorry for the confusion:
```
framework="tf": static graph (could be for both versions: tf1.x, but also tf2.x, if tf.compat.v1 mode is used)
framework="tf2": NOT static graph AND tf2.x version
```

**Comment by sven1977** [comment]
On the replay buffer question:
These additional memory tests run only `RolloutWorker.sample() + get_metrics()` on the rollout worker (both of which do not fill any buffers) or `Policy.learn_on_batch()` using a dummy batch, so no filling of the buffer either.

File: rllib/policy/tf_policy.py [hunk 16]
```diff
@@ -1187,11 +1187,12 @@ def _get_loss_inputs_dict(self, train_batch: SampleBatch, shuffle: bool):
         # Build the feed dict from the batch.
         feed_dict = {}
         for key, placeholders in self._loss_input_dict.items():
-            tree.map_structure(
+            a = tree.map_structure(
                 lambda ph, v: feed_dict.__setitem__(ph, v),
                 placeholders,
                 train_batch[key],
             )
+            del a
```
Comment (gjoliver) [inline rllib/policy/tf_policy.py:1195]: is this a memory leak, do we need to do this everywhere we use tree.map_structure?

**Review by gjoliver — APPROVED** [review]
